Fixed #25762 -- Refactored numberformat.format for speed

tim-mccurrach opened the pull request:
Optimises `numberformat.format` by utilising the bult-in `format` function where applicable. This is based off the work by Jaap Roes in [#5668](https://github.com/django/django/pull/5668).

See [https://code.djangoproject.com/ticket/25762](https://code.djangoproject.com/ticket/25762) for further details.

codingjoe commented on django/utils/numberformat.py:
```diff
@@ -66,6 +75,13 @@ def format(number, decimal_sep, decimal_pos=None, grouping=0, thousand_sep='',
     dec_part = dec_part and decimal_sep + dec_part
     # grouping
     if use_grouping:
+        # Use the builtin format where possible
```
Is there a way not to repeat the same code?

codingjoe commented on django/utils/numberformat.py:
```diff
@@ -1,7 +1,7 @@
+import builtins
```
You can also use `f-strings` down below. We only support Python versions that implement `f-strings`.

codingjoe commented on django/utils/numberformat.py:
```diff
@@ -22,12 +22,18 @@ def format(number, decimal_sep, decimal_pos=None, grouping=0, thousand_sep='',
-    # sign
-    sign = ''
-    if isinstance(number, Decimal):
+    if isinstance(number, int) and (not use_grouping or grouping == 3):
```
You're repeating `not grouping`, I am sure there is a way to rearrange the if statements a bit.

tim-mccurrach commented:
Hi @codingjoe 

Thanks for taking a look at this, sorry it took so long for me to get back to you. I started it at djangoCon 2019, and then normal work took over and it got forgotten about. Anyhow...

I agree this function is already pretty complex as it is, so I have tried to follow your suggestion to make it more readable. I think alot of the reason the function is tricky to follow, is that we have to deal with different types differently (for security/performance reasons). So I've split them up into sub-functions. I hope this makes the main function more readable, and the smaller functions single-purposed enough to be clear.

I think I've corrected all requested changes other than to re-introduce mark_safe. I'm not sure why this is needed. The current function only makes the returned value safe in one scenario (an integer, with no grouping, and no decimals - which wouldn't actually be escaped anyway). All other situations, it returns a normal string. It seemed sensible to just always return a normal string. I wondered if you had any further thoughts about this? (IMO it should be the responsibility of the template tags that call this function to decide if the returned value should be safe or not, but don't feel strongly about it.)

codingjoe commented on django/utils/numberformat.py:
```diff
@@ -18,46 +17,93 @@ def format(number, decimal_sep, decimal_pos=None, grouping=0, thousand_sep='',
+    """
+    if grouping and (number > 999 or number < -999):
+        # f'{number:,}' returns a string with a ',' as a thousand separator
+        number = f'{number:,}'
+        if thousand_sep != ',':
+            number = number.replace(',', thousand_sep)
```
IMHO I would simply use `_` not `,` and always replace. Reduces complexity and a replacement isn't too slow. In fact, it's probably faster than the if clause. If clauses can significantly slow down code execution, since your processor doesn't know which instructions to preload until you reached that particular instruction and decided where to go from there.

tim-mccurrach replied:
I've opted for `number = f'{number:,}'.replace(',', thousand_sep)` here:

Profiling the function: 
 - using `_` and removing the if clause led to a very slight increase in speed for the case when `thousand_sep != ','` (2-4% less time), but a larger decrease in speed for `thousand_sep == ','` (10-15% more time). 
 - This decrease in speed was reduced to (2-4% more time) when using a comma instead of `_`, and always replacing. (My guess is that `.replace` just does nothing if the arguments are equal?).

codingjoe commented on django/utils/numberformat.py:
```diff
@@ -18,46 +17,93 @@ def format(number, decimal_sep, decimal_pos=None, grouping=0, thousand_sep='',
+
+    This function only handles when grouping is 0 or 3.
+    """
+    if grouping and (number > 999 or number < -999):
```
IMHO the size check adds more complexity. Let python's format handle this. I believe it will very efficient if the value has only 3 digits.

tim-mccurrach replied:
I'm tempted to leave this one in. For numbers that are only 3 digits, it reduces the time taken by about 25%. I feel like thats enough to warrant the added complexity. I don't feel super-strongly about it though. 

(Obviously it does run faster on larger numbers, but the gains of not containing the check are small compared to the losses of formating/replacing all numbers).

codingjoe commented on django/utils/numberformat.py:
```diff
@@ -18,46 +17,93 @@ def format(number, decimal_sep, decimal_pos=None, grouping=0, thousand_sep='',
     else:
-        str_number = str(number)
+        number = str(number)
+    if decimal_pos:
+        return number + decimal_sep + '0' * decimal_pos
```
Why don't you do the decimal precision as part of the format expression? Something like:
```python
number = ("{:_.%sf}" % decimal_pos).format(number)
```

tim-mccurrach replied:
Using `format` to take care of the decimals here is alot slower. It's slower for both large and small, but for large numbers can take 2 or 3 times as long. (Which is a shame, it looks much nicer). So I've left it as it is.

pope1ni replied:
Or you can still use f-strings for this too:
```python
number = f'{number:_.{decimal_pos}f}'
```

tim-mccurrach replied:
@pope1ni This one seems to be slower as well. Especially as you then have to replace the decimal point when `decimal_sep` is something other than a full-stop/period.

codingjoe reviewed (commented):
Hi @tim-mccurrach,

I'm glad you found some time to continue your contribution here. I tried my hands on another review. I am afraid it might not be complete, since this is very hard to review. Maybe some tests would help especially since you split this into sub functions. Having unit tests might help to identify potential problems.
Best, Joe

PS: If you don't have too much time at your disposal, I would recommend keeping your changes to a minimum. You can always open another PR once this is merged.

tim-mccurrach commented on django/utils/numberformat.py:
```diff
@@ -18,46 +17,93 @@ def format(number, decimal_sep, decimal_pos=None, grouping=0, thousand_sep='',
-    use_grouping = (use_l10n or (use_l10n is None and settings.USE_L10N)) and settings.USE_THOUSAND_SEPARATOR
-    use_grouping = use_grouping or force_grouping
-    use_grouping = use_grouping and grouping != 0
+    if force_grouping:
+        pass
+    elif not settings.USE_THOUSAND_SEPARATOR:
+        grouping = 0
+    elif use_l10n is False:
+        grouping = 0
+    elif use_l10n is None and not settings.USE_L10N:
+        grouping = 0
```
The main speed-gain I was trying to achieve here was to avoid unnecessary checks (e.g. if `force_grouping` is true, or `USE_THOUSAND_SEPARATOR` is false, there is no point completing further checks). 

I agree that `use_grouping` is more explicit, and clearer. `grouping` would still only be a tuple or an integer. Since the default value of `grouping` is 0, I wasn't really thinking of it as a boolean (although I admit, we do treat it as a boolean when we say `if grouping:`). I removed `use_grouping` just because it was an extra variable to pass around. I tried adding use_grouping back in, and it didn't make much difference to speed (maybe 2% more time, but nothing worth fussing about). I could go either way on this.

pope1ni replied:
Perhaps you could use `grouping = None` instead of `grouping = 0`?

While not a more descriptive name, it is perhaps more semantically equivalent to grouping being disabled. Would that satisfy you, @codingjoe?

tim-mccurrach replied:
I certainly wouldn't have any objection to using `grouping=None`.

tim-mccurrach commented:
@codingjoe and @pope1ni . Thank you both very much for the valuable feedback. 

I think I've replied to or made changes in relation to all comments. Please do let me know if there's anything I can do to keep this moving.

jacobtylerwalls commented on django/utils/numberformat.py:
```diff
@@ -1,7 +1,7 @@
 from decimal import Decimal
 
 from django.conf import settings
-from django.utils.safestring import mark_safe
```
Hi @tim-mccurrach I'm inclined to agree, doesn't look like a major speed bump to revert this, since both `mark_safe` and `_format_int` have to evaluate a couple statements before casting to str. Your view on this?

jacobtylerwalls replied:
Ah, sorry, I see that your reply is found in the comment below this review. I see your view, but it seems to me this is a change in the underlying functionality, not just optimizing for speed. So it could slow down reviews/require a release note. Isn't it possible someone has code taking the result of numberformat.format and checking whether or not it's a SafeString before modifying it?

tim-mccurrach replied:
I've shaved a little bit off by returning a `SafeString` directly instead of calling `mark_safe`. It still is quite a bit slower for integers:
```
old_format:                       Mean +- std dev: 4.71 us +- 0.18 us
new_format(with safe strings):    Mean +- std dev: 4.30 us +- 0.22 us
new_format(without safe strings): Mean +- std dev: 3.11 us +- 0.12 us
``` 
Whilst it would be a breaking change, the current behaviour definitely seems like the 'wrong' behaviour. Nonetheless, in the interest of getting this ticket moving, I've made the change. It can always be addressed in a follow up ticket.

jacobtylerwalls commented on tests/utils_tests/test_numberformat.py:
```diff
@@ -1,35 +1,147 @@
+                {"USE_THOUSAND_SEPARATOR": False, "USE_L10N": True},
+                0
+            ),
+            (   # settings.USE_THOUSAND_SEPARATOR takes precendence over settings.USE_L10N
```
"precedence" here and above

jacobtylerwalls commented on tests/utils_tests/test_numberformat.py:
```diff
@@ -39,68 +151,56 @@ def test_large_number(self):
+            ('1234', '.', None, "", 0, '1234'),                           # Integer
+            ('1234.2', '.', None, " ", 0, '1234.2'),                      # With decimal
+            ('1234', '.', 2, ", ", 0, '1234.00'),                         # dec_pos specified
+            ('12345', ".", 2, ", ", 2, "1, 23, 45.00"),                   # dec_poa specified, and grouping
```
minor typo: dec_pos

carltongibson commented:
Hey @kezabelle — if you have a cycle/the interest, could I ask you to glance at this for initial comments? It's sat on the backlog for a while, and bit of fresh input might help us get it moving; feels like it's in your ballpark — do decline if not. 😊 Thanks!

kezabelle commented on django/utils/numberformat.py:
```diff
@@ -18,46 +18,99 @@ def format(number, decimal_sep, decimal_pos=None, grouping=0, thousand_sep='',
+        # It is unclear as to why this case should be marked safe, but
+        # we return SafeString here to avoid a change in functionality
+        # from the previous implementation of format.
+        number = SafeString(number)
```
I don't think you necessarily need this _here_ (if at all, but I get the historic reasoning) fwiw.

In the original (if I'm reading the side-by-side correctly) it ends up marked as safe when `dec_pos` is _Falsy_. But in this changed version, it becomes a `SafeString` _before_ checking `if decimal_pos`.

If there 's _truthy_ `decimal_pos` you end up doing `SafeString` + sep (presumably str, but could be another `SafeString`) + `str` so you ultimately get back a `str` type regardless, because `'0' * decimal_pos` is the final addition and `SafeString` only defines `__add__` for another of it's kind.

If you were able to defer making it a `SafeString` until _after_ the `decimal_pos` check, you can avoid 1 or possibly 2, accounting for sep) in the _wants decimal_ case, because `str` + (potentially) `SafeString` yields `str`.

NB: that's all off the top of my head from reading the diff, caveat emptor :)

kezabelle commented on django/utils/numberformat.py:
```diff
@@ -18,46 +18,99 @@ def format(number, decimal_sep, decimal_pos=None, grouping=0, thousand_sep='',
+
+    See numberformat.format docstring for details about arguments.
+    """
+    if decimal_pos is not None:
```
Here (and in `_format_string`) the check is `if decimal_pos is not None:` but previously in `_format_int`, and later in this function it's just `if decimal_pos:`

Perhaps should be consistent everywhere, and if not, a comment on _why_ not (could be a legitimate reason, but I can't discern it from glancing at it).

I _think_ `is not None` is every so slightly slower (understandably), but not so much much it matters in any tangible way ... probably.

kezabelle commented on django/utils/numberformat.py:
```diff
@@ -18,46 +18,99 @@ def format(number, decimal_sep, decimal_pos=None, grouping=0, thousand_sep='',
+        )
+        return f'{coefficient}e{exponent}'
+
+    number = f'{number:f}'
```
Could be a personal or project preference for all I know, but does this re-assignment need to occur, instead of just having `return _format_string(f'{number:f}', ...)`? Doubt it matters much, of course.

felixxm commented:
Closing due to inactivity.
